From 4980114fe868c62ee3ab8b0dba35d622580b6537 Mon Sep 17 00:00:00 2001
From: Tobias Gunkel <hennymcc@yahoo.de>
Date: Sat, 19 May 2018 16:14:11 -0400
Subject: [PATCH 09/11] gps.fix crash on access to unset
 AGpsRilCallbacks::request_refloc

Since Android 8.0 the AGpsRilCallbacks::request_refloc callback is not
initialized anymore. Before 8.0 the callback was initialized by
---
 gnss/1.0/default/Gnss.cpp | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/gnss/1.0/default/Gnss.cpp b/gnss/1.0/default/Gnss.cpp
index 32c131c8..4bad5be9 100644
--- a/gnss/1.0/default/Gnss.cpp
+++ b/gnss/1.0/default/Gnss.cpp
@@ -46,6 +46,25 @@ GpsCallbacks Gnss::sGnssCb = {
     .gnss_sv_status_cb = gnssSvStatusCb,
 };
 
+/*
+ * AGnssRilCallback implements the callback methods required by the AGnssRil
+ * interface.
+ */
+struct AGnssRilCallback : IAGnssRilCallback {
+    Return<void> requestSetIdCb(uint32_t setIdFlag) override;
+    Return<void> requestRefLocCb() override;
+};
+
+Return<void> AGnssRilCallback::requestSetIdCb(uint32_t setIdFlag) {
+    ALOGI("%s: called with setIdFlag=%d", __func__, setIdFlag);
+    return Void();
+}
+
+Return<void> AGnssRilCallback::requestRefLocCb() {
+    ALOGI("%s: called", __func__);
+    return Void();
+}
+
 uint32_t Gnss::sCapabilitiesCached = 0;
 uint16_t Gnss::sYearOfHwCached = 0;
 
@@ -511,6 +530,12 @@ Return<sp<IAGnssRil>> Gnss::getExtensionAGnssRil()  {
             ALOGI("%s: GnssRil interface not implemented by HAL", __func__);
         } else {
             mGnssRil = new AGnssRil(agpsRilIface);
+            sp<IAGnssRilCallback> aGnssRilCb = new AGnssRilCallback();
+            if (mGnssRil != nullptr) {
+                mGnssRil->setCallback(aGnssRilCb);
+            } else {
+                ALOGE("Unable to initialize AGnss Ril interface\n");
+            }
         }
     }
     return mGnssRil;
-- 
2.11.0

